presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,793 @@
+
+                blockEncodingBuffers = new BlockEncodingBuffers[channelCount];
+                for (int i = 0; i < channelCount; i++) {
+                    blockEncodingBuffers[i] = createBlockEncodingBuffers(decodedBlocks[i], initialElementCount);
+                }
+            }
+        }
+
+        private void calculateRowSizes(int fixedWidthRowSize, List<Integer> variableWidthChannels)
+        {
+            if (variableWidthChannels.isEmpty()) {
+                return;
+            }
+
+            ensureRowSizesCapacityAndInitialize();
+
+            for (int i : variableWidthChannels) {
+                blockEncodingBuffers[i].accumulateRowSizes(rowSizes);
+            }
+
+            for (int i = 0; i < positionCount; i++) {
+                rowSizes[i] += fixedWidthRowSize;
+            }
+        }
+
+        private int calculateNextBatchSize(int fixedWidthRowSize, List<Integer> variableWidthChannels, int startPosition)
+        {
+            int bytesRemaining = capacity - toIntExact(getSerializedBuffersSizeInBytes());
+
+            if (variableWidthChannels.isEmpty()) {
+                int positionsFit = max(bytesRemaining / fixedWidthRowSize, 1);
+                bufferFull = (positionsFit <= positionCount - startPosition);
+                return min(positionsFit, positionCount - startPosition);
+            }
+
+            verify(rowSizes != null);
+            for (int i = startPosition; i < positionCount; i++) {
+                bytesRemaining -= rowSizes[i];
+
+                if (bytesRemaining <= 0) {
+                    bufferFull = true;
+                    return max(i - startPosition, 1);
+                }
+            }
+
+            return positionCount - startPosition;
+        }
+
+        private void flush(OutputBuffer outputBuffer)
+        {
+            if (bufferedRowCount == 0) {
+                return;
+            }
+
+            SliceOutput output = new DynamicSliceOutput(toIntExact(getSerializedBuffersSizeInBytes()));
+            output.writeInt(channelCount);
+
+            for (int i = 0; i < channelCount; i++) {
+                blockEncodingBuffers[i].serializeTo(output);
+            }
+
+            SerializedPage serializedPage = serde.serialize(output.slice(), bufferedRowCount);
+            outputBuffer.enqueue(lifespan, partition, ImmutableList.of(serializedPage));
+            pagesAdded.incrementAndGet();
+            rowsAdded.addAndGet(bufferedRowCount);
+
+            if (blockEncodingBuffers != null) {
+                for (int i = 0; i < blockEncodingBuffers.length; i++) {
+                    blockEncodingBuffers[i].resetBuffers();
+                }
+            }
+
+            bufferedRowCount = 0;
+        }
+
+        private void ensurePositionsCapacity(int positionCount)
+        {
+            if (positions.length < positionCount) {
+                positions = new int[positionCount];
```

Comment:
this logic is inconsistent with similar logic elsewhere in this file; other places use `max(2 * positions.length, positionCount)`